### ISSUE TITLE
Add a middleware to modify query parameters

### ISSUE THREAD

**Issue by ldez** [containous/traefik#6276]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

Around the same idea of the [ReplacePath](https://docs.traefik.io/v2.1/middlewares/replacepath/)/[ReplacePathRegex](https://docs.traefik.io/v2.1/middlewares/replacepathregex/) middlewares.

Goals:

- use query parameters to modify the query parameters
  ex: `/foo?a=1&b=2&c=3` -> `/foo?a=1&b=2&c=6` or `/foo?a=1&b=2&c=3&d=4`
- use query parameters to modify the path
  ex: `/foo?a=one&b=two&c=three` -> `/foo/one?a=one&b=two&c=three`
- use query parameters to modify the path and the query parameters
  ex: `/foo?a=one&b=two&c=three` -> `/foo/one?b=two&c=three` or `/foo/a/one?b=two&c=three` or `/foo/one/two/three`
- use path to modify the query parameters
  ex: `/foo/a/one/b/two/c/three` -> `/foo?a=one&b=two&c=three` or `/foo/one/two/three` -> `/foo?a=one&b=two&c=three`
- use path to modify the path and the query parameters
  ex: `/foo/d/four?b=two&c=three` -> `/foo?b=two&c=three&d=four` or `/foo/b/four?b=two&c=three` -> `/foo?b=four&c=three`

Note:

The regular expressions cannot really use in those cases as I explained here: https://github.com/containous/traefik/pull/4820#issuecomment-552195957

A query parameter can have multiple values `/foo?a=1&a=10&a=2&a=3`

**Comment by tolgaakyuz** [comment]
hey @ldez.

We are currently adapting traefik into our internal infrastructure tooling cluster. This feature is something we would like to see happening in the near future. 

Our use case is simple, we are running an image manipulation server that accepts data via `path` such as `/path/to/some/image/1200px/other/things`. As you can see the part of the path (`1200px`) is dynamic and it's currently coming from a query parameter, like so `?size=1200px`.  The full translation is something like this `example.com/path/to/image?size=1200px` -> `example.com/path/to/image/1200px`.

I would like to pick this up and try to come up with a solution that covers the use cases that you defined in the description. Let me know if this would be an interest of yours as well.

**Comment by ldez** [comment]
@tolgaakyuz feel free to create a PR, take a look to #4820 in order not to use the same approach.

**Comment by ptruman** [comment]
Definitely have use cases for this.

One example is a (so far) not updated web app I'm using behind Traefik which wants to put a password in the querystring (it does not use POST as yet) - and the easiest way to do that is to update the querystring to add ?password=XXXXX - which would allow access AND hide the password from the end user/browser - however replacepath/replacepathregex don't work.

**Comment by nemanjan00** [comment]
#6994 while closed deffinitly is a bug and needs to be fixed. At this point, there is no way to rewrite url to move something from query to path.

**Comment by modem7** [comment]
Would definitely be useful to remove the facebook fbclid from URLs, It can break far too many things.

**Comment by karimodm** [comment]
Hey guys! A collegue and me have just discovered the "path only" limitation of `ReplacePathRegex` middleware in our favorite proxy software and came across this issue.
Just want to express my support for this feature and bump it up a little!

**Comment by zambotn** [comment]
Yes, this would be needed in at least a couple of projects: those services now have an additional Express.js server that performs this replacement but this is just a dirty hack.

**Comment by ldez** [comment]
You can use a plugin: https://pilot.traefik.io/plugins/280116342026142217/query-paramter-modification

**Comment by rarecrumb** [comment]
> You can use a plugin: https://pilot.traefik.io/plugins/280116342026142217/query-paramter-modification

I dont think this plugin works... I'm just trying add query parameters in every request to a defined service. Is there any way?

**Comment by deleted** [comment]
I could really use this feature as well

My use case is that I'm runniing an instance of PostgREST to automatically generate a REST API against a postgres schema, and I'd like it to answer requests with the route `/:resource_name/:resource_id`. The postgREST's author's recommended means of doing so is to rewrite the URL so that the `:resource_id` parameter from the path winds up in the query string. This is possible with nginx as shown here: https://postgrest.org/en/stable/explanations/nginx.html?highlight=alternate%20url#alternate-url-structure

I'd rather not run a second reverse proxy, so It would be great if this were possible in traefik as well.

**Comment by Konubinix** [comment]
Hi. I just came to the same issue. I wanted to replace the url `/ipfs/<some-hash>/somefile.pdf` into `/pdfviewer?file=<some-hash>/somefile.pdf` and found out I could not.

So, I'm BUMPing this as well

**Comment by twbworld** [comment]
我遇到同样问题,无法实现:
`/abc.html` 
修改为
`/index?u=abc`

**Comment by fer-ri** [comment]
I want to replace `2023/07/26/wr-v.jpg` with `/?url=2023/07/26/wr-v.jpg`. But Its not working.

Bump  :+1:

**Comment by lc-guy** [comment]
Is there any progress to be eventually made on this?
It's quite a basic feature, and using any of the replacePath middlewares results in failure due to `?`'s and `&`'s being escaped by traefik into %3F and %26, respectively. All of the associated tickets related to fixing this behavior were also closed while redirecting to this ticket, which has been open for 3 years at this point.
I really don't mean to sound entitled with regards to it, but a solution should be proposed at any rate, no?

**Comment by geraldcroes** [comment]
Hi there, 

we discussed it during triage, and believe this would prove a nice addition to Traefik indeed.

We'd love to see an incoming PR to accelerate the timeline (hence the contributor/wanted label we just added)

**Comment by ct-condat** [comment]
Is an update in sight in the meantime? We currently always have to switch to Nginx elsewhere in order to carry out such redirects.

**Comment by g-roliveira** [comment]
### Installing the Traefik Plugin and Configuring for Query Parameters and Path Modification

After facing challenges with handling query parameters and path rewriting in Traefik, I successfully resolved my issue using the `traefik-plugin-query-modification`. Here are the details on how I configured everything, from installing the plugin to the final configuration that worked for me.

**1. Installing the Plugin in Traefik:**

First, I needed to activate the plugin in Traefik. This involved adding the following lines to the Traefik configuration file to ensure the plugin was loaded correctly:

```shell
--experimental.plugins.traefik-plugin-query-modification.modulename=github.com/kingjan1999/traefik-plugin-query-modification
--experimental.plugins.traefik-plugin-query-modification.version=v1.0.0
```

These settings enable the experimental plugin, specifying the module name and the version of the plugin.

**2. Configuring the Middleware in Docker Compose:**

After installing the plugin, I configured the necessary middleware in my `docker-compose.yml`. This included middleware to replace the path and another to add the required query parameter:

```yaml
services:
  # Middleware to replace path
  - "traefik.http.middlewares.replace-path.replacepathregex.regex=^/webhook/app$$"
  - "traefik.http.middlewares.replace-path.replacepathregex.replacement=/webhook-app-notifica.rule"

  # Middleware to add query parameter
  - "traefik.http.middlewares.add-query.plugin.traefik-plugin-query-modification.type=add"
  - "traefik.http.middlewares.add-query.plugin.traefik-plugin-query-modification.paramName=sys"
  - "traefik.http.middlewares.add-query.plugin.traefik-plugin-query-modification.newValue=APP"
  - "traefik.http.routers.apprewrite.middlewares=replace-path,add-query"
```

This configuration allowed transforming requests coming to `/webhook/app` into `/webhook-app-notifica.rule?sys=APP`. The `replace-path` middleware modifies the access path while the `add-query` adds the `sys=APP` query parameter.

**Final Result:**

With this setup, requests aimed at the endpoint `/webhook/app` are now correctly redirected to `/webhook-app-notifica.rule` with the query parameter `sys=APP` included, exactly as needed. The server log shows that the final path received is `/webhook-app-notifica.rule?sys=APP`, indicating that the rewriting and addition of parameters worked as expected.

I hope this explanation helps others who are facing similar challenges with Traefik and need an effective solution for handling query parameters and URL paths.

**Comment by s1495k043** [comment]
I don't need new middleware, I just want it to not encode my question marks
